euphonic/qpoint_phonon_modes.py

```diff
@@ -390,6 +395,123 @@ def calculate_debye_waller(
+            or cross_sections are supplied, the neutron-weighted DOS is
+            returned and y-axis is in area/energy units per atom, where the
+            first spectrum is the total neutron-weighted DOS per average
+            atom, and the other spectra show the contributions of each
+            species to the total
+        """
+        weighting_opts = [None, 'coherent', 'incoherent']
+        if not weighting in weighting_opts:
+            raise ValueError(f'Invalid value for weighting, got '
+                             f'{weighting}, should be one of '
+                             f'{weighting_opts}')
+
+        spec_idx_dict = self.crystal.get_species_idx()
+        if isinstance(cross_sections, str) and weighting is not None:
+            cross_sections_data = get_reference_data(
+                collection=cross_sections,
+                physical_property=f'{weighting}_cross_section')
+        elif isinstance(cross_sections, dict):
+            cross_sections_data = cross_sections
+        else:
+            cross_sections_data = None
+        if cross_sections_data is not None:
+            cs = [cross_sections_data[x] for x in spec_idx_dict.keys()]
+            # Account for cross sections in different, or invalid, units
+            ex_units = '[length]**2'
+            if not cs[0].check(ex_units):
+                raise ValueError((
+                    f'Unexpected dimensionality in cross_sections units, '
+                    f'expected {ex_units}, got {str(cs[0].dimensionality)}'))
+            cs_unit = cs[0].units
+            cs = [x.to(cs_unit).magnitude for x in cs]*cs_unit
+        else:
+            cs = None
+
+        evec_weights = np.real(np.einsum('ijkl,ijkl->ijk',
+                                         self.eigenvectors,
+                                         np.conj(self.eigenvectors)))
+        n_modes = self._frequencies.shape[1]
+
+        for i, (spec, spec_idx) in enumerate(spec_idx_dict.items()):
+            species_weights = np.sum(
+                evec_weights[:, :, spec_idx], axis=-1)
+            dos = self._calculate_dos(dos_bins, mode_widths=mode_widths,
+                                       mode_widths_min=mode_widths_min,
+                                       mode_weights=species_weights)
+            if cs is not None:
+                # Neutron weighted DOS is per atom of sample
+                avg_atom_mass = np.mean(self.crystal._atom_mass)
+                species_mass = self.crystal._atom_mass[spec_idx[0]]
+                dos *= cs[i]*avg_atom_mass/species_mass
+            if i == 0:
+                all_dos_y_data = np.zeros((
+                    len(spec_idx_dict) + 1, len(dos_bins) - 1))*dos.units
+            all_dos_y_data[i + 1] = dos
+
+        all_dos_y_data[0] = np.sum(all_dos_y_data[1:], axis=0)
+        metadata = {'line_data':
+            [{'label': x} for x in ['Total'] + list(spec_idx_dict.keys())]}
```

Comment:
This seems to rely on `dict.keys()` having a consistent order with `spec_idx_dict.items()`. Having the base dict() object be consistently ordered is

- an implementation detail of CPython 3.6
- part of the language specification from Python 3.7

So in principle it could break if somebody uses e.g. pypy 3.6 or IronPython 3.6. (IronPython only just caught up with version 3.4, so this is not totally farfetched.)

A cheap and robust fix would be to modify `Crystal.get_species_idx` to use `collections.OrderedDict`. This is a no-op for CPython and makes the intent to _use_ the ordering explicit.

Reply:
This did occur to me at one point but then I forgot... thanks, fixed!